README.md
RWListFieldCalculations
=======================

A simple add-on to enable the use of List fields in calculations.

Requires dev build of Gravity Forms 1.9

NOT FOR PRODUCTION USE

Installation
============

1. Download ZIP and install via WordPress admin or upload plugin to the /wp-content/plugins/ directory
2. Activate the plugin through the 'Plugins' menu in the WordPress Admin Panel
3. Using the merge tag drop down above the calculation formula textarea you can add merge tags for list fields to the formula.  Merge tags will be available for counting the list field rows and the individual column sums.


changelog.txt
-------------------------------------------------------------------------------------------------------------------
Version 0.7
    - Use gravity form hook gform_loaded for loading the addon

-------------------------------------------------------------------------------------------------------------------
Version 0.6
    - Updated min Gravity Forms version to 2.3
    - Fixed the gform_merge_tag_value_pre_calculation callback for Gravity Forms 2.4.

-------------------------------------------------------------------------------------------------------------------
Version 0.5
    - Updated to use object notation for the $field instead of array access.
    - Updated min Gravity Forms version to 1.9.9.8
    - Fixed typo in the gform_post_calculation_events function.

-------------------------------------------------------------------------------------------------------------------
Version 0.4
    - LFCalc.js replaced gform_calculation_result with two new hooks: 
      - gform_calculation_event to handle triggering calculation when row added/removed and on colummn input change event
      - gform_calculation_merge_tag_value to replace merge tag with row count or column total
    - Remove function hooked to gform_get_form_filter
    - increased minimum version to Gravity Forms 1.9
    -------------------------------------------------------------------------------------------------------------------
Version 0.3
    - Changed columnSelector from input to :input to add support for drop down enabled columns
    - Changed on method from blur event to change event to add support for drop downs
-------------------------------------------------------------------------------------------------------------------
Version 0.2
    - Switched to gform_get_form_filter from gform_pre_render so inline script is positioned after the from
    - Switched to gform_enqueue_scripts so script is only enqueued if a list field merge tag is found
-------------------------------------------------------------------------------------------------------------------
Version 0.1
    - Added List Field merge tags to the calculations merge tag dropdown
    - Moved to add-on framwework


js/LFCalc.js
gform.addAction( 'gform_post_calculation_events', function( mergeTagArr, formulaField, formId, calcObj ){
    var fieldId = parseInt( mergeTagArr[1] ),
        fieldSelector = '#field_' + formId + '_' + fieldId;

    if ( jQuery( fieldSelector + ' table.gfield_list' ).length == 1 ) {
        jQuery( fieldSelector )
            .on( 'click', '.add_list_item', function () {
                jQuery( fieldSelector + ' .delete_list_item' ).removeProp( 'onclick' );
                calcObj.bindCalcEvent( fieldId, formulaField, formId, 0 );
            })
            .on( 'click', '.delete_list_item', function () {
                gformDeleteListItem( this, 0 );
                calcObj.bindCalcEvent( fieldId, formulaField, formId, 0 );
            });

        if ( mergeTagArr[2] != null ) {
            var columnNo = mergeTagArr[2].substr( 1 ),
                columnSelector = '.gfield_list_' + fieldId + '_cell' + columnNo + ' :input';
            jQuery( fieldSelector ).on( 'keyup', columnSelector, function () {
                calcObj.bindCalcEvent( fieldId, formulaField, formId, 0 );
            });
        }
    }
});

gform.addFilter( 'gform_merge_tag_value_pre_calculation', function( value, mergeTagArr, isVisible, formulaField, formId ){
    var fieldId = parseInt( mergeTagArr[1] ),
        fieldSelector = '#field_' + formId + '_' + fieldId;
    
    // check if merge tag belongs to a List field and that it isn't hidden by conditional logic
    if ( jQuery( fieldSelector + ' table.gfield_list' ).length == 1 && isVisible ) {

        if ( mergeTagArr[2] == null ) {
            // if no column specified count the rows
            value = jQuery( fieldSelector + ' table.gfield_list tbody tr.gfield_list_group' ).length;
        } else {
            var columnNo = mergeTagArr[2].substr( 1 ),
                columnSelector = '.gfield_list_' + fieldId + '_cell' + columnNo + ' :input',
                cellValue = 0;

            value = 0;
                
            // if column specified get the input values from each row and calculate the sum
            jQuery( columnSelector ).each( function () {
                cellValue = gformToNumber( jQuery( this ).val() );
                value += parseFloat( cellValue ) || 0;
            });
        }

    }

    return value;
});


RWListFieldCalculations.php
<?php
/*
Plugin Name: Gravity Forms List Field Calculations Add-On
Plugin URI:
Description: A simple add-on to enable the use of List fields in calculations.
Version: 0.7
Author: Richard Wawrzyniak
Author URI:

------------------------------------------------------------------------
Copyright 2014 Richard Wawrzyniak

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
*/

//------------------------------------------

add_action( 'gform_loaded', function(){
    GFForms::include_addon_framework();

    class RWListFieldCalculations extends GFAddOn {

        protected $_version = '0.7';
        protected $_min_gravityforms_version = '2.3';
        protected $_slug = 'RWListFieldCalculations';
        protected $_path = 'RWListFieldCalculations/RWListFieldCalculations.php';
        protected $_full_path = __FILE__;
        protected $_title = 'Gravity Forms List Field Calculations Add-On';
        protected $_short_title = 'List Field Calculations';

        public function init() {
            parent::init();
            add_action( 'gform_enqueue_scripts', array( $this, 'list_field_calculations_script' ), 10, 2 );
            add_filter( 'gform_calculation_formula', array( $this, 'list_field_calculations' ), 10, 4 );
            add_filter( 'gform_custom_merge_tags', array( $this, 'list_field_calculations_merge_tags' ), 10, 4 );
        }

        public static function has_list_field_merge_tag( $form ) {
            foreach ( $form['fields'] as $field ) {
                if ( ! $field->has_calculation() ) {
                    continue;
                }

                preg_match_all( '/{[^{]*?:(\d+)\.?(\d+)?}/mi', $field->calculationFormula, $matches, PREG_SET_ORDER );

                if ( is_array( $matches ) ) {

                    foreach ( $matches as $match ) {

                        // get the $field object for the provided id
                        $field_id = $match[1];
                        $lfield   = GFAPI::get_field( $form, $field_id );

                        // check that we have a field
                        if ( ! $lfield ) {
                            continue;
                        }

                        // check the field type as we only want the rest of the function to run if the field type is list
                        if ( $lfield->get_input_type() != 'list' ) {
                            continue;
                        }

                        return true;

                    }

                }

            }

            return false;
        }

        public function list_field_calculations_script( $form ) {

            if ( self::has_list_field_merge_tag( $form ) ) {
                wp_enqueue_script( 'LFCalc', $this->get_base_url() . '/js/LFCalc.js', array(
                    'jquery',
                    'gform_gravityforms'
                ), $this->_version, true );
            }

        }

        public function list_field_calculations( $formula, $field, $form, $lead ) {

            // {List:1.3} - {Label:ID.Column} - sum column values
            // {List:1} - {Label:ID} - count rows
            preg_match_all( '/{[^{]*?:(\d+)\.?(\d+)?}/mi', $formula, $matches, PREG_SET_ORDER );

            if ( is_array( $matches ) ) {

                foreach ( $matches as $match ) {

                    // $match[0] = merge tag e.g. {List:1.3}
                    // $match[1] = field id e.g. 1
                    // $match[2] = column number e.g. 3

                    // get the $field object for the provided id
                    $field_id = $match[1];
                    $field    = GFAPI::get_field( $form, $field_id );

                    // check that we have a field
                    if ( ! $field ) {
                        continue;
                    }

                    // check the field type as we only want the rest of the function to run if the field type is list
                    if ( $field->get_input_type() != 'list' ) {
                        continue;
                    }

                    // get the list fields values from the $lead
                    $list_values = empty( $lead[ $field_id ] ) ? array() : unserialize( $lead[ $field_id ] );
                    $count       = 0;

                    // if column number found sum column values otherwise count number of rows
                    if ( isset( $match[2] ) ) {

                        // count the actual number of columns
                        $column_count = count( $field->choices );

                        if ( $column_count > 1 ) {
                            // subtract 1 from column number as the choices array is zero based
                            $column_num = $match[2] - 1;
                            // get the column label so we can use that as the key to the multi-column values
                            $column = rgars( $field->choices, "{$column_num}/text" );
                        }

                        foreach ( $list_values as $value ) {
                            if ( $column_count == 1 ) {
                                $count += GFCommon::to_number( $value );
                            } else {
                                $count += GFCommon::to_number( $value[ $column ] );
                            }
                        }

                    } else {
                        $count = count( $list_values );
                    }

                    // replace the merge tag with the row count or column sum
                    $formula = str_replace( $match[0], $count, $formula );

                }

            }

            // return the modified formula so the rest of the calculation can be run by GF
            return $formula;
        }

        public function list_field_calculations_merge_tags( $merge_tags, $form_id, $fields, $element_id ) {

            // check the type of merge tag dropdown
            if ( $element_id != 'field_calculation_formula' ) {
                return $merge_tags;
            }

            foreach ( $fields as $field ) {

                // check the field type as we only want to generate merge tags for list fields
                if ( $field->get_input_type() != 'list' ) {
                    continue;
                }

                $label = $field->label;
                $tag = '{' . $label . ':' . $field->id;
                $column_count = count( $field->choices );

                if ( $column_count > 1 ) {

                    $i = 0;

                    foreach ( $field->choices as $column ) {
                        $merge_tags[] = array(
                            'label' => $label . ' - ' . $column['text'] . ' (Column sum)',
                            'tag'   => $tag . '.' . ++ $i . '}'
                        );
                    }

                } else {
                    $merge_tags[] = array( 'label' => $label . ' (Column sum)', 'tag' => $tag . '.1}' );
                }

                $merge_tags[] = array( 'label' => $label . ' (Row count)', 'tag' => $tag . '}' );

            }

            return $merge_tags;
        }

    }

    new RWListFieldCalculations();
}, 10, 0 );
